File: jamii/graph/interaction_graph.py
"""
Interaction Graph — in-memory User-to-Post interaction graph.
Inspired by Twitter's GraphJet / UTEG (User-Tweet Entity Graph).

Maintains bidirectional edges:
  user → posts they engaged with
  post → users who engaged with it

Used for out-of-network discovery via graph traversal:
"Find posts that users similar to me have engaged with."
"""

from typing import Dict, Set, List, Tuple
from collections import defaultdict
import random


class InteractionGraph:
    def __init__(self):
        self._user_to_posts: Dict[str, Set[str]] = defaultdict(set)
        self._post_to_users: Dict[str, Set[str]] = defaultdict(set)

    def add_edge(self, user_id: str, post_id: str) -> None:
        self._user_to_posts[user_id].add(post_id)
        self._post_to_users[post_id].add(user_id)

    def posts_for_user(self, user_id: str) -> Set[str]:
        return self._user_to_posts.get(user_id, set())

    def users_for_post(self, post_id: str) -> Set[str]:
        return self._post_to_users.get(post_id, set())

    def discover_via_traversal(
        self,
        viewer_id: str,
        viewer_following: List[str],
        exclude_post_ids: Set[str],
        depth: int = 2,
        max_results: int = 50,
    ) -> List[str]:
        """
        Traverse: viewer's follows → their liked posts → other users who liked those posts
        → other posts those users liked. Returns discovered post_ids.
        """
        visited_users: Set[str] = {viewer_id}
        discovered_posts: Set[str] = set()

        seed_users = set(viewer_following) & set(self._user_to_posts.keys())
        if not seed_users:
            return []

        level_1_posts: Set[str] = set()
        for uid in seed_users:
            level_1_posts.update(self._user_to_posts.get(uid, set()))
        level_1_posts -= exclude_post_ids

        if depth == 1:
            return list(level_1_posts)[:max_results]

        similar_users: Set[str] = set()
        for pid in level_1_posts:
            engagers = self._post_to_users.get(pid, set())
            similar_users.update(engagers - visited_users)
        visited_users.update(similar_users)

        for uid in similar_users:
            posts = self._user_to_posts.get(uid, set())
            discovered_posts.update(posts - exclude_post_ids - level_1_posts)

        result = list(discovered_posts)
        random.shuffle(result)
        return result[:max_results]

    def total_edges(self) -> int:
        return sum(len(posts) for posts in self._user_to_posts.values())

File: java/graphjet/InteractionGraph.java
package com.blusce.jamii.graphjet;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * InteractionGraph — in-memory User-to-Post interaction graph.
 *
 * Directly mirrors Twitter's GraphJet (UTEG — User-Tweet Entity Graph):
 * maintains a real-time, in-memory bipartite graph between users and posts,
 * enabling fast out-of-network discovery via graph traversal.
 *
 * Reference: twitter/the-algorithm → src/scala/com/twitter/recos/
 *            https://github.com/twitter/GraphJet
 *
 * Graph structure:
 *   Left nodes:  Users
 *   Right nodes: Posts
 *   Edges:       User → Post (weighted by engagement type)
 *
 * Discovery algorithm (2-hop traversal):
 *   1. Start from viewer's followed accounts
 *   2. Find posts they engaged with (Level 1)
 *   3. Find other users who engaged with those posts (Level 2)
 *   4. Collect posts from those users (Level 3)
 *   → Returns ranked list of discovered post IDs
 */
public class InteractionGraph {

    // Engagement weights matching Twitter's open-source formula
    public static final Map<String, Double> ENGAGEMENT_WEIGHTS = Map.of(
        "like",              0.5,
        "retweet",           1.0,
        "reply",            13.5,
        "bookmark",          2.0,
        "profile_click",    12.0,
        "negative",        -74.0,
        "report",         -369.0
    );

    // Adjacency: user_id → { post_id → engagement_weight }
    private final ConcurrentHashMap<String, Map<String, Double>> userToPost
        = new ConcurrentHashMap<>();

    // Reverse adjacency: post_id → { user_id → engagement_weight }
    private final ConcurrentHashMap<String, Map<String, Double>> postToUser
        = new ConcurrentHashMap<>();

    // Total edge count (for monitoring)
    private volatile long edgeCount = 0;

    // ─── Mutation ─────────────────────────────────────────────────────────────

    /**
     * Add a weighted engagement edge.
     * engagementType: "like" | "retweet" | "reply" | "bookmark" | "report" etc.
     */
    public void addEdge(String userId, String postId, String engagementType) {
        double weight = ENGAGEMENT_WEIGHTS.getOrDefault(engagementType, 0.1);
        if (weight <= 0) return;  // Don't index negative signals as edges

        userToPost
            .computeIfAbsent(userId, k -> new ConcurrentHashMap<>())
            .merge(postId, weight, Double::sum);

        postToUser
            .computeIfAbsent(postId, k -> new ConcurrentHashMap<>())
            .merge(userId, weight, Double::sum);

        edgeCount++;
    }

    public void addEdge(String userId, String postId) {
        addEdge(userId, postId, "like");
    }

    // ─── Query ────────────────────────────────────────────────────────────────

    public Map<String, Double> getPostsForUser(String userId) {
        return userToPost.getOrDefault(userId, Collections.emptyMap());
    }

    public Map<String, Double> getUsersForPost(String postId) {
        return postToUser.getOrDefault(postId, Collections.emptyMap());
    }

    /**
     * Discover out-of-network posts via 2-hop traversal.
     *
     * @param viewerId         the user requesting the feed
     * @param viewerFollowing  accounts the viewer follows
     * @param excludePostIds   posts already seen or in feed
     * @param maxResults       max discovered posts to return
     * @return ranked list of (postId, score) pairs
     */
    public List<Map.Entry<String, Double>> discover(
        String viewerId,
        Set<String> viewerFollowing,
        Set<String> excludePostIds,
        int maxResults
    ) {
        // Level 1: posts liked by viewer's follows
        Map<String, Double> level1Posts = new HashMap<>();
        for (String followedId : viewerFollowing) {
            Map<String, Double> theirPosts = userToPost.getOrDefault(followedId, Collections.emptyMap());
            for (Map.Entry<String, Double> e : theirPosts.entrySet()) {
                if (!excludePostIds.contains(e.getKey())) {
                    level1Posts.merge(e.getKey(), e.getValue(), Double::sum);
                }
            }
        }

        // Level 2: similar users (who also liked Level 1 posts)
        Map<String, Double> similarUsers = new HashMap<>();
        for (Map.Entry<String, Double> postEntry : level1Posts.entrySet()) {
            Map<String, Double> engagers = postToUser.getOrDefault(postEntry.getKey(), Collections.emptyMap());
            for (Map.Entry<String, Double> engager : engagers.entrySet()) {
                String uid = engager.getKey();
                if (!viewerFollowing.contains(uid) && !uid.equals(viewerId)) {
                    similarUsers.merge(uid, engager.getValue() * postEntry.getValue(), Double::sum);
                }
            }
        }

        // Level 3: posts from similar users (not already seen)
        Map<String, Double> discovered = new HashMap<>();
        for (Map.Entry<String, Double> userEntry : similarUsers.entrySet()) {
            Map<String, Double> theirPosts = userToPost.getOrDefault(userEntry.getKey(), Collections.emptyMap());
            for (Map.Entry<String, Double> postEntry : theirPosts.entrySet()) {
                String postId = postEntry.getKey();
                if (!excludePostIds.contains(postId) && !level1Posts.containsKey(postId)) {
                    double score = postEntry.getValue() * userEntry.getValue();
                    discovered.merge(postId, score, Double::sum);
                }
            }
        }

        // Sort by score descending
        return discovered.entrySet().stream()
            .sorted(Map.Entry.<String, Double>comparingByValue().reversed())
            .limit(maxResults)
            .collect(Collectors.toList());
    }

    // ─── Stats ────────────────────────────────────────────────────────────────

    public long getTotalEdges() { return edgeCount; }
    public int getUserCount()   { return userToPost.size(); }
    public int getPostCount()   { return postToUser.size(); }

    @Override
    public String toString() {
        return String.format("InteractionGraph{users=%d, posts=%d, edges=%d}",
            getUserCount(), getPostCount(), getTotalEdges());
    }
}
